### ISSUE TITLE
Moize cache is mixed up and produces wrong result in a recursive function

### ISSUE THREAD

**Issue by zuyetawarmatik** [planttheidea/moize#95]
Hi,

Thanks for the lib that has been working well so far in our project.
I'm having an issue with memoize function from moize (I tried micro-memoize, memoizee or fast-memoize, but they all have no problem in this case).

The function to be tested is `aggregateData`. Typically it will aggregate a parent `data` Object with attr `children` as Array.

Example (SUM agg on `a` and `b` attributes): 
Input object:
```js
{
  children: [
    { children: [{ a: 1, b: 1 }] },
    { children: [{ a: 1, b: 1 }] }
  ]
}
```
Result object:
```js
{
  children: [
    { children: [{ a: 1, b: 1 }], a: { agg: {SUM: 1} }, b: { agg: {SUM: 1} } },
    { children: [{ a: 1, b: 1 }], a: { agg: {SUM: 1} }, b: { agg: {SUM: 1} } }
  ], 
  a: { agg: {SUM: 2} },
  b: { agg: {SUM: 2} }
}
```

Source
```js
import { get, set, union, omitBy, isNil, isEmpty } from 'lodash';
import memoize from 'moize';

const AGG_TYPE = {
  INCLUSIVE: 'INCLUSIVE',
  SUM: 'SUM',
  AVG: 'AVG',
  COUNT: 'COUNT'
};

const { INCLUSIVE, SUM, AVG, COUNT } = AGG_TYPE;

export const getFieldValue = (item, field) => {
  let value = typeof field === 'string' ? item[field] : get(item, field);
  if (typeof value === 'object' && value !== null) value = value.value;
  return value;
};

export const setFieldValue = (item, field, value) => {
  if (typeof field === 'string') item[field] = value;
  else set(item, field, value);
};

export const getFieldAggValue = (item, field) => {
  const value = typeof field === 'string' ? item[field] : get(item, field);
  if (typeof value === 'object' && value !== null) return value.agg;
  return undefined;
};

export const unionInclusiveObject = (objA, objB) => {
  if (isNil(objA) && isNil(objB)) {
    return undefined;
  } else if (isNil(objA) && !isNil(objB)) {
    return typeof objB === 'object' ? { ...objB } : undefined;
  } else if (!isNil(objA) && isNil(objB)) {
    return typeof objA === 'object' ? { ...objA } : undefined;
  } else if (typeof objA !== 'object' || typeof objB !== 'object') {
    return undefined;
  }

  const sum = {};
  const objAKeys = Object.keys(objA);
  const objBKeys = Object.keys(objB);
  const keys = union(objAKeys, objBKeys);
  keys.forEach((key) => {
    sum[key] = (objA[key] || 0) + (objB[key] || 0);
  });

  return sum;
};

export const sumWithNil = (a, b) => {
  if (isNil(a) && isNil(b)) return undefined;
  else if (isNil(a) && !isNil(b)) return b;
  else if (!isNil(a) && isNil(b)) return a;
  return a + b;
};

export const aggregateData = memoize((data, aggMetadata) => {
  const AGG_NAMES = Object.keys(aggMetadata);
  const { children } = data;
  if (isEmpty(children)) return data;

  // Recursion
  const newChildren = children.map(row =>
    (row.children ? aggregateData(row, aggMetadata) : row)
  );

  const aggData = {};

  AGG_NAMES.forEach((aggName) => {
    const { ref: field = aggName, type, rounding } = aggMetadata[aggName];

    let inclusiveAgg;
    let sumAgg;
    let avgAgg;
    let countAgg;

    newChildren.forEach((row) => {
      const rowAggValue = getFieldAggValue(row, field);
      const rowValue = getFieldValue(row, field);

      // Determine if it aggregates on aggregated value or on primitive value
      // Aggregate on aggregated: group row aggregated from group rows
      // Aggregate on primitive: group row aggregated from element rows

      const isAggOnAgg = !isNil(rowAggValue);
      const canAggOnPri = !isNil(rowValue);

      if (isAggOnAgg) {
        countAgg = sumWithNil(countAgg, rowAggValue[COUNT]);
      } else if (canAggOnPri) {
        countAgg = sumWithNil(countAgg, 1);
      }

      if (type === INCLUSIVE) {
        if (isAggOnAgg) {
          inclusiveAgg = unionInclusiveObject(inclusiveAgg, rowAggValue[INCLUSIVE]);
        } else if (canAggOnPri) {
          inclusiveAgg = unionInclusiveObject(inclusiveAgg, { [rowValue]: 1 });
        }
      } else if (type === SUM || type === AVG) {
        if (isAggOnAgg) {
          sumAgg = sumWithNil(sumAgg, rowAggValue[SUM]);
        } else if (canAggOnPri) {
          sumAgg = sumWithNil(sumAgg, rowValue);
        }
      }
    });

    if (type === AVG && countAgg && !isNil(sumAgg)) {
      avgAgg = sumAgg / countAgg;
      if (rounding) {
        const { strategy } = rounding;
        if (strategy === 'truncate') avgAgg = Math.trunc(avgAgg);
      }
    }

    setFieldValue(aggData, field, {
      agg: omitBy({
        [INCLUSIVE]: inclusiveAgg,
        [SUM]: sumAgg,
        [AVG]: avgAgg,
        [COUNT]: countAgg
      }, isNil)
    });
  });

  return {
    ...data,
    ...aggData,
    children: newChildren
  };
});
```

Test:

If you run aggregateData once, the test is passed. aggregateData also works with 2-level data. But for 3-level data, if you run twice, the key-value of the memoize cache is mismatched.

Everything is OK if I use other memoize libs. Somehow there's problem in the cache update with moize in this case.

```js
import { expect } from 'chai';
import { aggregateData } from './dataHelper';
describe('3-level data', () => {
  const aggMetadata = {
    a: { type: COUNT },
    b: { type: SUM },
    c: { type: AVG },
    d: { type: INCLUSIVE }
  };
  const el1 = { a: 'sku_1', b: 1, c: 2, d: 'cat_1' };
  const el2 = { a: 'sku_2', b: 3, c: 4, d: 'cat_1' };
  const el3 = { a: 'sku_3', b: 5, c: undefined, d: 'cat_2' };
  const el4 = { a: 'sku_4', b: null, c: 0, d: 'cat_2' };
  const el5 = { a: 'sku_5', b: undefined, c: null, d: 'cat_2' };

  it('aggregate on data without aggregated value yet', () => {
    const data = {
      children: [{ children: [el1, el2] }, { children: [el3, el4, el5] }]
    };
    const aggData1 = aggregateData(data, aggMetadata);
    const aggData2 = aggregateData(data, aggMetadata);
    expect(aggData1).to.deep.equal(aggData2);
  });
});
```

**Comment by planttheidea** [comment]
There is a lot going on here ... I can take a look, but if you could possibly distill this down to a simpler example that reproduces the issue, that would help tremendously.

Also, can you explain what you mean by mismatched? As in, what is the expected result, what is the actual result.

**Comment by zuyetawarmatik** [comment]
Hi, you can refer to the example on the beginning of my post. The problem is only for the aggregateData function. Other functions in the snippet are just dependency for test reproduction.

Basically:
- Given that the memoized `aggregateData()` aggregates on an object `x`, the expected output is `x_agg`.
- Given that `x` contains 2 `children`, which is `x1` and `x2`: `x.children = [x1, x2]`.
- Given that `x1` contains 2 `children`: `x1.children = [x11, x12]`.
- Given that `x2` contains 2 `children`: `x2.children = [x21, x22]`.
- Then, `aggregateData` will recursively calculate aggregated values on `x`, `x1`, and `x2`. So there will be 3 cached results: `aggregateData(x)`, `aggregateData(x1)`, `aggregateData(x2)`.

This is OK if I run memoized `aggregateData` once (by using `moize`); and OK if using other libraries, included `micro-memoize`, for any number of runs.
But it will give incorrect result if I run `aggregateData` twice and more than that (using `moize`):

After 1st time:
```
// OK result
agregateData(x) = x_agg;
agregateData(x1) = x1_agg;
agregateData(x2) = x2_agg;
```

After 2nd time:
```
// Mismatched cache, there's a swap of agregateData(x) and agregateData(x1)
agregateData(x) = x1_agg;
agregateData(x1) = x_agg;
agregateData(x2) = x2_agg;
```

I'm quite sure there's problem with cache update of `moize`, as the other libs for me work OK with my algorithm.

**Comment by planttheidea** [comment]
Oh I'm not trying to minimize it, I just wanted a bit more detail for my investigation. Let me play with it and see what I find.

**Comment by planttheidea** [comment]
So good news and bad news ... the good news is that I was able to reproduce a much simpler example to reproduce the issue. The issue is actually `micro-memoize` under the hood ... the reason you weren't seeing any issues is because `micro-memoize` by default has a cache size of 1, whereas `moize` has an infinite cache by default. There is indeed something hinky going on with the cache updates, where the `keys` and `values` with the cache are getting out of alignment.

I will be investigating this further, but for now a few potential workarounds you can use:
- use `moize.simple` instead of `moize` ... this sets the cache size to 1
- instead of recursively calling the memoized function, store the function in a variable and call that recursively

```javascript
const fn = (data, aggMetadata) => {
  const AGG_NAMES = Object.keys(aggMetadata);
  const { children } = data;
  if (isEmpty(children)) return data;

  // Recursion
  const newChildren = children.map(row =>
    (row.children ? fn(row, aggMetadata) : row)
  );

  const aggData = {};

  AGG_NAMES.forEach((aggName) => {
    const { ref: field = aggName, type, rounding } = aggMetadata[aggName];

    let inclusiveAgg;
    let sumAgg;
    let avgAgg;
    let countAgg;

    newChildren.forEach((row) => {
      const rowAggValue = getFieldAggValue(row, field);
      const rowValue = getFieldValue(row, field);

      // Determine if it aggregates on aggregated value or on primitive value
      // Aggregate on aggregated: group row aggregated from group rows
      // Aggregate on primitive: group row aggregated from element rows

      const isAggOnAgg = !isNil(rowAggValue);
      const canAggOnPri = !isNil(rowValue);

      if (isAggOnAgg) {
        countAgg = sumWithNil(countAgg, rowAggValue[COUNT]);
      } else if (canAggOnPri) {
        countAgg = sumWithNil(countAgg, 1);
      }

      if (type === INCLUSIVE) {
        if (isAggOnAgg) {
          inclusiveAgg = unionInclusiveObject(inclusiveAgg, rowAggValue[INCLUSIVE]);
        } else if (canAggOnPri) {
          inclusiveAgg = unionInclusiveObject(inclusiveAgg, { [rowValue]: 1 });
        }
      } else if (type === SUM || type === AVG) {
        if (isAggOnAgg) {
          sumAgg = sumWithNil(sumAgg, rowAggValue[SUM]);
        } else if (canAggOnPri) {
          sumAgg = sumWithNil(sumAgg, rowValue);
        }
      }
    });

    if (type === AVG && countAgg && !isNil(sumAgg)) {
      avgAgg = sumAgg / countAgg;
      if (rounding) {
        const { strategy } = rounding;
        if (strategy === 'truncate') avgAgg = Math.trunc(avgAgg);
      }
    }

    setFieldValue(aggData, field, {
      agg: omitBy({
        [INCLUSIVE]: inclusiveAgg,
        [SUM]: sumAgg,
        [AVG]: avgAgg,
        [COUNT]: countAgg
      }, isNil)
    });
  });

  return {
    ...data,
    ...aggData,
    children: newChildren
  };
};

export const aggregateData = moize(fn);
```

I'll work on this in `micro-memoize` and report back findings here, but that's the best I got for now.

**Comment by planttheidea** [comment]
Released the fix to `micro-memoize` as version `2.0.4`, and I released a new version of `moize` that includes the bump in dependencies as `5.3.2`. I checked both with my simple reproduction and also with your specific use-case, both passed as expected, so I think you should be good to go.

If you have any other issues, let me know!
